### PR TITLE
fp16 conv

### PR DESCRIPTION
**Pull request by chenfucn**
### Description

Convolution for fp16 datatype.  Use NHWC for computation. For NCHW input, it rearranges the input tensor to NHWC format before computing the result.

Support two optional fusion:

1. Activation
2. Add (not yet implemented)

### Motivation and Context

Accelerating fp16 inference

### CODE REVIEW

File: onnxruntime/core/providers/cpu/fp16/fp16_conv.cc [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+  bool channels_last_{false};
+  TensorShape W_shape_;
+  BufferUniquePtr packed_W_buffer_;
+  size_t packed_W_size_{0};
+  bool is_W_packed_{false};
+  BufferUniquePtr reordered_W_buffer_;
+};
+
+
+Status FusedConvFp16::PrePack(const Tensor& tensor, int input_idx, AllocatorPtr alloc,
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/fp16/fp16_conv.cc:97]: Can you please add a test like this - https://github.com/microsoft/onnxruntime/blob/32533dd1c290aa6025c584a705d3fa00fde1cd0a/onnxruntime/test/providers/cpu/math/matmul_test.cc#L297 to ensure that sharing of prepacked weights at a session level is kicking in ?

Comment (chenfucn) [reply onnxruntime/core/providers/cpu/fp16/fp16_conv.cc:101]: added

File: onnxruntime/core/providers/cpu/fp16/fp16_conv.cc [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+ * The implementation runs faster with NHWC. By default, it converts NCHW to NHWC
+ * before processing, and convert the result back. It can take NHWC tensors directly.
+ * Use operator attribute 'channels_last' to specify that the data layout is NHWC.
+ *
+*/
+class FusedConvFp16 final : public OpKernel {
+ public:
+  FusedConvFp16(const OpKernelInfo& info) : OpKernel(info), conv_attrs_(info) {
+    ORT_ENFORCE(GetFusedActivationAttr(info, activation_).IsOK());
+    channels_last_ = (info.GetAttrOrDefault<int64_t>("channels_last", static_cast<int64_t>(0)) != 0);
```
Comment (yufenglee) [inline onnxruntime/core/providers/cpu/fp16/fp16_conv.cc:46]: >channels_last [](http://example.com/codeflow?start=54&length=13)

Neither Conv nor FusedConv have channel_last attribute. We need to define a Contrib op to support NHWC for Conv fp16.

File: onnxruntime/core/mlas/inc/mlas.h
Comment (jchen351) [inline onnxruntime/core/mlas/inc/mlas.h:1]: NIT: Consider to split this file into smaller files? and have mlas.h to include {smaller_mlas1.h smaller_mlas2.h...} just to improve the readability

**Review by jchen351 — APPROVED** [review]

File: onnxruntime/core/providers/cpu/fp16/fp16_conv.cc (hunk 2 above)
Comment (chenfucn) [reply onnxruntime/core/providers/cpu/fp16/fp16_conv.cc:46]: fix it in later PR

**Pull request merged**